crabs/detection_tracking/models.py

```diff
@@ -1,83 +1,216 @@
 
 
 class FasterRCNN(LightningModule):
     """
     LightningModule implementation of Faster R-CNN for object detection.
 
-    Parameters:
-    -----------
+    Parameters
+    ----------
     config : dict
         Configuration settings for the model.
 
-    Methods:
-    --------
+    Methods
+    -------
     forward(x):
         Forward pass of the model.
     training_step(batch, batch_idx):
         Defines the training step for the model.
     configure_optimizers():
         Configures the optimizer for training.
 
-    Attributes:
-    -----------
+    Attributes
+    ----------
     config : dict
         Configuration settings for the model.
     model : torch.nn.Module
         Faster R-CNN model.
+    training_step_outputs : dict
+        Dictionary to store training metrics.
+    validation_step_outputs : dict
+        Dictionary to store validation metrics.
+    test_step_outputs : dict
+        Dictionary to store test metrics.
     """
 
-    def __init__(self, config):
+    def __init__(self, config: Dict[str, Any]):
         super().__init__()
         self.config = config
-        self.model = torchvision.models.detection.fasterrcnn_resnet50_fpn_v2(
-            weights=torchvision.models.detection.FasterRCNN_ResNet50_FPN_V2_Weights.DEFAULT
-        )
-        in_features = self.model.roi_heads.box_predictor.cls_score.in_features
-        self.model.roi_heads.box_predictor = FastRCNNPredictor(
-            in_features, config["num_classes"]
-        )
+        self.model = self.configure_model()
         self.training_step_outputs = {
-            "total_training_loss": 0.0,
+            "training_loss_epoch": 0.0,
+            "num_batches": 0,
+        }
+        self.validation_step_outputs = {
+            "precision_epoch": 0.0,
+            "recall_epoch": 0.0,
             "num_batches": 0,
         }
+        self.test_step_outputs = {
+            "precision_epoch": 0.0,
+            "recall_epoch": 0.0,
+            "num_batches": 0,
+        }
+
+    def configure_model(self) -> torch.nn.Module:
+        """
+        Configures the Faster R-CNN model with default weights,
+        specified backbone, and box predictor.
+        """
+        model = fasterrcnn_resnet50_fpn_v2(weights="DEFAULT")
+        in_features = model.roi_heads.box_predictor.cls_score.in_features
+        model.roi_heads.box_predictor = faster_rcnn.FastRCNNPredictor(
+            in_features, self.config["num_classes"]
+        )
+        return model
 
-    def forward(self, x):
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Forward pass of the model.
+        """
         return self.model(x)
 
-    def training_step(self, batch, batch_idx):
-        images, targets = batch
-        loss_dict = self.model(images, targets)
-        total_loss = sum(loss for loss in loss_dict.values())
+    def accumulate_metrics(
+        self,
+        outputs: List[Dict[str, Union[float, int]]],
```

Comment:
when we use this function, isn't outputs always a 1-element list? should we then simplify this function?